File: app/src/main/java/com/kacper/and_krakgo/screens/sign_in/reset_password/ResetPasswordFragment.java
package com.kacper.and_krakgo.screens.sign_in.reset_password;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.TextInputLayout;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;

import com.kacper.and_krakgo.InputTypes;
import com.kacper.and_krakgo.R;
import com.kacper.and_krakgo.helpers.ToastMessageHelper;
import com.kacper.and_krakgo.helpers.ValidationHelper;
import com.kacper.and_krakgo.rxbus.SignInEvents;
import com.kacper.and_krakgo.views.CustomTextWatcher;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.reactivex.functions.Consumer;

/**
 * Created by kacper on 05/11/2017.
 */

public class ResetPasswordFragment extends Fragment implements ResetPasswordContract.View {

    @BindView(R.id.reset_email_input) TextInputLayout mResetEmail;
    @BindView(R.id.progress_bar) ProgressBar mProgressBar;

    private ResetPasswordPresenter mPresenter;
    public static final String TAG = "ResetPasswordFragment";

    private boolean isLoading = false;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mPresenter = new ResetPasswordPresenter(this);
        SignInEvents.subscribe(new Consumer<Integer>() {
            @Override
            public void accept(Integer o) throws Exception {
                if (isVisible() && o == SignInEvents.ACTION_BOTTOM_BUTTON_PASSWORD) {
                    isLoading = true;
                    showLoading();
                    resetPassword();
                }
            }
        });
    }

    private void resetPassword() {
        if (mResetEmail.getEditText() != null && mResetEmail.isErrorEnabled()) {
            ToastMessageHelper.showShortToast(R.string.error_email_field);
        } else {
            mPresenter.sendResetPassword(mResetEmail.getEditText().getText().toString());
        }
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_reset_password, container, false);
        ButterKnife.bind(this, rootView);
        setListeners();
        return rootView;
    }

    private void setListeners() {
        mResetEmail.getEditText().addTextChangedListener(new CustomTextWatcher() {
            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                super.onTextChanged(charSequence, i, i1, i2);
                String error = ValidationHelper.validateText(getContext(), charSequence.toString(), InputTypes.EMAIL);
                if (error == null) {
                    mResetEmail.setErrorEnabled(false);
                } else {
                    mResetEmail.setEnabled(true);
                    mResetEmail.setError(error);
                }
            }
        });
    }

    @Override
    public void showError(String message) {
        ToastMessageHelper.showShortToast( message);
        isLoading = false;
        showLoading();
    }

    @Override
    public void onResetSuccesful() {
        ToastMessageHelper.showShortToast( R.string.password_change_success);
        getActivity().onBackPressed();
    }

    private void showLoading() {
        if (isLoading) {
            mProgressBar.setVisibility(View.VISIBLE);
        } else {
            mProgressBar.setVisibility(View.GONE);
        }
        mResetEmail.setEnabled(!isLoading);
    }


}

File: app/src/main/java/com/kacper/krakgo/screens/sign_in/reset_password/ResetPasswordFragment.kt
package com.kacper.krakgo.screens.sign_in.reset_password

import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.kacper.krakgo.model.enums.InputTypes
import com.kacper.krakgo.R
import com.kacper.krakgo.helpers.ToastMessageHelper
import com.kacper.krakgo.helpers.ValidationHelper
import com.kacper.krakgo.rxbus.SignInEvents
import com.kacper.krakgo.views.CustomTextWatcher

import butterknife.ButterKnife
import com.kacper.krakgo.mvp.MvpFragment
import kotlinx.android.synthetic.main.fragment_reset_password.*

/**
 * Created by kacper on 05/11/2017.
 */

class ResetPasswordFragment : MvpFragment<ResetPasswordContract.View,
        ResetPasswordContract.Presenter>(),
        ResetPasswordContract.View {

    override var mPresenter: ResetPasswordContract.Presenter = ResetPasswordPresenter()

    private var isLoading = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        SignInEvents.subscribe { o ->
            if (isVisible && o == SignInEvents.ACTION_BOTTOM_BUTTON_PASSWORD) {
                isLoading = true
                showLoading()
                resetPassword()
            }
        }
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        setListeners()
    }

    private fun resetPassword() {
        if (reset_email_input.editText != null && reset_email_input.isErrorEnabled) {
            ToastMessageHelper.showShortToast(context, R.string.error_email_field)
        } else {
            mPresenter.sendResetPassword(reset_email_input.editText?.text.toString())
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val rootView = inflater.inflate(R.layout.fragment_reset_password, container, false)
        ButterKnife.bind(this, rootView)
        return rootView
    }

    private fun setListeners() {
        reset_email_input.editText?.addTextChangedListener(object : CustomTextWatcher() {
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                super.onTextChanged(charSequence, i, i1, i2)
                val error = ValidationHelper.validateText(context, charSequence.toString(), InputTypes.EMAIL)
                if (error == null) {
                    reset_email_input?.isErrorEnabled = false
                } else {
                    reset_email_input?.isEnabled = true
                    reset_email_input?.error = error
                }
            }
        })
    }

    override fun showError(error: String) {
        ToastMessageHelper.showShortToast(context, error)
        isLoading = false
        showLoading()
    }

    override fun onResetSuccessful() {
        ToastMessageHelper.showShortToast(context, R.string.password_change_success)
        activity?.onBackPressed()
    }

    private fun showLoading() {
        if (isLoading) {
            progress_bar.visibility = View.VISIBLE
        } else {
            progress_bar.visibility = View.GONE
        }
        reset_email_input.isEnabled = !isLoading
    }

    companion object {
        const val TAG = "ResetPasswordFragment"
    }


}
